lib/aida/skills/decision_tree.ex

```diff
@@ -0,0 +1,210 @@
+defmodule Aida.Skill.DecisionTree do
+  alias __MODULE__
+  alias Aida.{Bot, Message, Message.TextContent, Session, Skill.Utils}
+  alias Aida.Skill.DecisionTree.{Question, Answer, Response}
+
+  @type t :: %__MODULE__{
+    id: String.t,
+    bot_id: String.t,
+    name: String.t,
+    explanation: Bot.message,
+    clarification: Bot.message,
+    keywords: %{},
+    relevant: nil | Aida.Expr.t,
+    tree: [Question.t | Answer.t]
+  }
+
+  defstruct id: "",
+            bot_id: "",
+            name: "",
+            explanation: %{},
+            clarification: %{},
+            keywords: %{},
+            relevant: nil,
+            tree: %{}
+
+  defmodule Question do
+    @type t :: %__MODULE__{
+      id: String.t,
+      question: Aida.Bot.message,
+      responses: [Response.t]
+    }
+
+    defstruct id: "",
+              question: %{},
+              responses: []
+
+    def valid_answer?(question, message) do
+      response = find_response(question, message)
+      case response do
+        nil -> false
+        _ -> true
+      end
+    end
+
+    def next_question_id(question, message) do
+      response = find_response(question, message)
+      case response do
+        nil -> question.id
+        _ -> response.next
+      end
+    end
+
+    defp find_response(question, message) do
+      question.responses |> Enum.find(&(Enum.any?(Map.values(&1.keywords), fn(x) -> Enum.member?(x, String.downcase(Message.text_content(message))) end)))
```

Comment:
I'd split this a little bit. Maybe delegating some of the code to the `Response` module.
Something like:
```elixir
message = Message.text_content(message) |> String.downcase
question.responses |> Enum.find(&Response.matches?(message))
```